tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,354 @@
+    """
+    is_running = is_container_running(duthost, container_name)
+    if is_running == should_be_running:
+        return True
+    else:
+        return False
+
+
+def get_program_status(duthost, container_name, program_name):
+    """
+    @summary: Return the status of a program in the specified container
+    @return: "RUNNING" or "EXITING" represents the program is in the running
+             or exited status
+    """
+    program_status = None
+
+    process_list = duthost.shell("docker exec {} supervisorctl status".format(container_name))
+    for process_info in process_list["stdout_lines"]:
+        if process_info.find(program_name) != -1:
+            program_status = process_info.split()[1].strip()
+            break
+
+    return program_status
+
+
+def kill_process_by_pid(duthost, container_name, program_name, program_pid):
+    """
+    @summary: Kill a process in the specified container by its pid
+    """
+    duthost.shell("docker exec {} kill -SIGKILL {}".format(container_name, program_pid))
+
+    # Wait 30 seconds such that the container is stopped
+    wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+               CONTAINER_CHECK_INTERVAL_SECS,
+               check_container_state, duthost, container_name, False)
+
+    is_running = is_container_running(duthost, container_name)
+    if is_running:
+        program_status = get_program_status(duthost, container_name, program_name)
+        # If program is still in the running state, that means it is not terminated by "kill" command
+        # and the testing will exit
+        if program_status == "RUNNING":
+            pytest.fail("Failed to stop {} process before test".format(program_name))
+
+    logger.info("{} in {} container is stopped successfully".format(program_name, container_name))
+
+
+def is_hiting_start_limit(duthost, container_name):
+    """
+    @summary: Determine whether the container can not be restarted is due to
+              start-limit-hit or not
+    """
+    service_status = duthost.shell("sudo systemctl status {}.service | grep 'Active'".format(container_name))
+    for line in service_status["stdout_lines"]:
+        if "start-limit-hit" in line:
+            return True
+
+    return False
+
+
+def verify_autorestart_with_critical_process(duthost, container_name, program_name,
+                                             program_status, program_pid):
+    """
+    @summary: Killing a critical process in a container to verify whether the container
+              can be stopped and then restarted correctly
+    """
+    if program_status == "RUNNING":
+        kill_process_by_pid(duthost, container_name, program_name, program_pid)
+    elif program_status in ["EXITED", "STOPPED", "STARTING"]:
+        pytest.fail("Process '{}' in container {} is in the {} state, expected 'RUNNING'"
+                    .format(program_name, container_name, program_status))
+    else:
+        pytest.fail("Failed to find {} in {} container"
```

Comment:
Enhance message: `"Failed to find program '{}' in container '{}'"` Also applies to other similar messages.

Reply:
Updated!